### PR TITLE
Make CertificateData.ReadReverseRdns() accept T61String.

### PR DESCRIPTION
**Pull request by baulig**
`T61String` support has been added to `DerSequenceReader` in https://github.com/dotnet/corefx/pull/30572, but not to the managed reader in `CertificateData`.

We have a test for this in Mono:
https://github.com/mono/mono/blob/master/mcs/class/System/Test/System.Security.Cryptography.X509Certificates/X509Certificate2Test.cs - the `Certificate_2_Properties` fails in the call to `x.GetNameInfo (X509NameType.SimpleName, true)` on line 543.

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs [hunk 1]
```diff
@@ -471,6 +471,9 @@ private static string FindAltNameMatch(byte[] extensionBytes, GeneralNameType ma
                         case DerSequenceReader.DerTag.UTF8String:
                             value = tavReader.ReadUtf8String();
                             break;
+                        case DerSequenceReader.DerTag.T61String:
+                            value = tavReader.ReadT61String ();
```
Comment (danmosemsft) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs]: nit: space before parens. I know it's Mono style :disappointed_relieved:

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs [hunk 2]
```diff
@@ -471,6 +471,9 @@ private static string FindAltNameMatch(byte[] extensionBytes, GeneralNameType ma
                         case DerSequenceReader.DerTag.UTF8String:
                             value = tavReader.ReadUtf8String();
                             break;
+                        case DerSequenceReader.DerTag.T61String:
+                            value = tavReader.ReadT61String();
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs]: A test should be added for this new code.  Preferably one which doesn't require dual licensing.  If nothing else, it should be relatively easy to craft a certificate using the CertificateRequest API.

**Comment by baulig** [comment]
Added a test.  For future reference, this is how this is done.

1. Create `t61-string-cert.conf` containing

```
[req]
prompt = no
string_mask = nombstr # make sure this matches the value in openssl.conf
distinguished_name = server_name
string_mask = MASK:0x0004
encrypt_key = no

[server_name]
commonName = test-server.local
countryName = US
stateOrProvinceName = Massachusetts
organizationName = Xamarin
organizationalUnitName = Engineering
emailAddress = mabaul@microsoft.com

[server_exts]
keyUsage = critical,keyEncipherment,digitalSignature
extendedKeyUsage = critical,serverAuth
```

2. Generate the certificate with

```
openssl req -x509 -config t61string-cert.conf -days 3650 -newkey rsa:512 -keyout t61string-cert.key -out t61string-cert.pem
```

3. Generate hexdump with

```
hexdump -v -e '32/1 "%02x" "\n"'  t61string-cert.pem
```

4. Copy and paste the output into `TestData.cs` and add quotation / newlines.

File: src/System.Security.Cryptography.X509Certificates/tests/CertTests.cs [hunk 3]
```diff
@@ -363,6 +363,43 @@ public static void ExportPublicKeyAsPkcs12()
...
+            using (var cert = new X509Certificate2(TestData.T61StringCertificate))
+            {
+                Assert.Equal(certSubject, cert.Subject);
+                Assert.Equal(certSubject, cert.Issuer);
+
+                int snlen = cert.GetSerialNumber().Length;
+                Assert.Equal(9, snlen);
+
+                byte[] serialNumber = new byte[snlen];
+                Buffer.BlockCopy(cert.GetSerialNumber(), 0, serialNumber, 0, snlen);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/CertTests.cs]: GetSerialNumber returns a new byte[].  Why make another one and copy it?

File: src/System.Security.Cryptography.X509Certificates/tests/CertTests.cs [hunk 4]
```diff
@@ -363,6 +363,43 @@ public static void ExportPublicKeyAsPkcs12()
...
+
+                Assert.Equal(0x9e, serialNumber[0]);
+                Assert.Equal(0x7a, serialNumber[1]);
+                Assert.Equal(0x5c, serialNumber[2]);
+                Assert.Equal(0xcc, serialNumber[3]);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/tests/CertTests.cs]: Assert.Equal("9E7A5CCC...", serialNumber.ByteArrayToHex()); is both more compact and (I think) more traceable (in particular, it shows the context of what went wrong, for endianness, or padding byte, or whatever).

**Pull request merged**